### PR TITLE
Test(e2e): Modal Visual Tests #DS-1247

### PR DESCRIPTION
**Pull request by pavelklibani**
### Issue reference

[Udělat vizuální test otevřeného Modalu](https://jira.almacareer.tech/browse/DS-1247)

### CODE REVIEW

File: packages/web/src/scss/components/Modal/index.html [hunk 1]
```diff
@@ -303,26 +305,28 @@ <h2 id="example-dropdown-title" class="ModalHeader__title">Modal Title</h2>
             provident unde. Eveniet, iste, molestiae?
           </p>
           <div class="DropdownWrapper">
```
Comment (crishpeen) [inline packages/web/src/scss/components/Modal/index.html:307]: Can you please remove this as well? AFAIK is does nothing

File: tests/e2e/demo-components-compare.spec.ts-snapshots/fileuploader-chromium-linux.png
Comment (crishpeen) [inline tests/e2e/demo-components-compare.spec.ts-snapshots/fileuploader-chromium-linux.png:1]: 🤔

File: tests/e2e/demo-homepages.spec.ts-snapshots/web-chromium-linux.png
Comment (crishpeen) [inline tests/e2e/demo-homepages.spec.ts-snapshots/web-chromium-linux.png:1]: this is probably wrong, there should be the toggle

File: tests/e2e/demo-homepages.spec.ts-snapshots/web-react-chromium-linux.png
Comment (crishpeen) [inline tests/e2e/demo-homepages.spec.ts-snapshots/web-react-chromium-linux.png:1]: again, probably old screenshot. Please regenerate them.

File: tests/e2e/demo-modal-compare.spec.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+const runComponentCompareTests = ({ componentsDir, packageName, componentName }) => {
+  if (!packageName) return;
+
+  const formattedPackageName = packageName
+    .split('-')
+    .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
+    .join(' ');
+
+  test.describe(`Demo ${formattedPackageName} Components`, () => {
```
Comment (crishpeen) [inline tests/e2e/demo-modal-compare.spec.ts:14]: The names are not correct, are they?

File: tests/e2e/demo-modal-compare.spec.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+  // Helper function for taking and comparing screenshots
+  const takeScreenshot = async (suffix, fullPage = false) => {
+    await expect(page).toHaveScreenshot(`${componentName.toLowerCase()}-${suffix}.png`, {
+      animations: 'disabled',
+      fullPage,
+    });
+  };
+
+  // TEST: open basic modal, close with backdrop click
+  await page.click('[data-test-id="basic-modal"]');
```
Comment (crishpeen) [inline tests/e2e/demo-modal-compare.spec.ts:57]: I wonder if we should somehow mention that we do these tests in the demos, so developers know they can't modify them freely. What do you think?

File: tests/e2e/demo-modal-compare.spec.ts
Comment (crishpeen) [inline tests/e2e/demo-modal-compare.spec.ts:1]: Maybe we should place this in some directly where more component tests can live?

File: tests/e2e/demo-modal-compare.spec.ts [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    .split('-')
+    .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
+    .join(' ');
+
+  test.describe(`Demo ${formattedPackageName} Components`, () => {
+    test(`test demo ${formattedPackageName} component ${componentName}`, async ({ page }) => {
+      try {
+        const url = isTestingEnvironment()
+          ? SERVERS.TESTING[packageName]
+          : getDevelopmentEndpointUri(packageName, { isDocker: packageName !== 'web-twig' });
```
Comment (crishpeen) [inline tests/e2e/demo-modal-compare.spec.ts:19]: A lot of the same code as elsewhere, so maybe we should abstract it somehow?

File: tests/e2e/demo-modal-compare.spec.ts [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+  // TEST: open modal with hidden close button, close with action button
+  await page.click('button[data-test-id="hidden-close-button"]');
+  await takeScreenshot('hidden-close-button');
+  await page.click('dialog[open] .ModalFooter button');
+  await takeScreenshot('closed-modals', true);
+};
+
+const testConfigs = [
+  {
+    componentName: 'Modal',
```
Comment (crishpeen) [inline tests/e2e/demo-modal-compare.spec.ts:83]: do we need this?

File: packages/web-react/src/components/Modal/demo/ModalDefault.tsx [hunk 6]
```diff
@@ -25,7 +25,7 @@ const ModalDefault = () => {
   const [footerAlign, setFooterAlign] = useState<AlignmentXDictionaryType>(AlignmentX.RIGHT);
   const [isDockedOnMobile, setIsDockedOnMobile] = useState(false);
   const [isDropdownOpen, setIsDropdownOpen] = useState(false);
-  const [isExpandedOnMobile, setIsExpandedOnMobile] = useState(false);
+  const [isExpandedOnMobile, setIsExpandedOnMobile] = useState(true);
```
Comment (crishpeen) [inline packages/web-react/src/components/Modal/demo/ModalDefault.tsx:28]: this was inconsistent?

**Review by crishpeen — COMMENTED** [review]
Great work, just some ideas :)

File: packages/web-react/src/components/Modal/demo/ModalDefault.tsx (hunk 6 above)
Comment (pavelklibani) [reply packages/web-react/src/components/Modal/demo/ModalDefault.tsx:28]: yes, it was checked in web and twig, so i changed it in react to match other packages

File: tests/e2e/demo-homepages.spec.ts-snapshots/web-chromium-linux.png
Comment (pavelklibani) [reply tests/e2e/demo-homepages.spec.ts-snapshots/web-chromium-linux.png:1]: updated

File: tests/e2e/demo-components-compare.spec.ts-snapshots/fileuploader-chromium-linux.png
Comment (pavelklibani) [reply tests/e2e/demo-components-compare.spec.ts-snapshots/fileuploader-chromium-linux.png:1]: fixed, removed

File: tests/e2e/demo-modal-compare.spec.ts (hunk 4 above)
Comment (pavelklibani) [reply tests/e2e/demo-modal-compare.spec.ts]: made some helper functions

File: tests/e2e/demo-modal-compare.spec.ts
Comment (pavelklibani) [reply tests/e2e/demo-modal-compare.spec.ts:1]: moved to components folder

File: tests/e2e/demo-modal-compare.spec.ts (hunk 3 above)
Comment (pavelklibani) [reply tests/e2e/demo-modal-compare.spec.ts]: Hmm, I'm not sure how to handle this... I don't want to add comments all over the code. Personally, I think the data attribute is self-explanatory and indicates that it's for testing purposes. If someone changes or deletes it, the tests will alert them. 😄

File: tests/e2e/demo-modal-compare.spec.ts (hunk 5 above)
Comment (pavelklibani) [reply tests/e2e/demo-modal-compare.spec.ts]: I at least put it in a constant so the string doesn't repeat.

File: tests/e2e/components/demo-modal-compare.spec.ts [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        await runModalTests(page, componentName);
+      } catch (error) {
+        console.error(`Test for demo ${formattedPackageName} component ${componentName} failed. ${error}`);
+        throw error;
+      }
+    });
+  });
+};
+
+const runModalTests = async (page: Page, componentName: string): Promise<void> => {
```
Comment (curdaj) [inline tests/e2e/components/demo-modal-compare.spec.ts:32]:
```suggestion
type ModalTestConfig = {
  openSelector: string;
  testName: string;
};

const modalTestConfigs: ModalTestConfig[] = [
  {
    openSelector: 'modal-with-form',
    testName: 'with-form',
  },
  {
    openSelector: 'modal-with-dropdown',
    testName: 'with-dropdown',
  },
  {
    openSelector: 'modal-with-long-content',
    testName: 'with-long-content',
  },
  {
    openSelector: 'modal-with-scrolling-inside',
    testName: 'with-scrolling-inside',
  },
  {
    openSelector: 'modal-with-scrollview',
    testName: 'with-scrollview',
  },
  {
    openSelector: 'modal-with-custom-height',
    testName: 'with-custom-height',
  },
  {
    openSelector: 'modal-with-disabled-backdrop-click',
    testName: 'disabled-backdrop-click',
  },
];

const runModalTests = async (page: Page, componentName: string): Promise<void> => {
  for (const config of modalTestConfigs) {
    await page.click(`[data-test-id="${config.openSelector}"]`);
    await takeScreenshot(page, `${componentName}-${config.testName}`);
    await page.click('dialog[open] header button');
  }
  
    // open basic modal, close with backdrop click
  await page.click('[data-test-id="modal-basic"]');
  await takeScreenshot(page, `${componentName}-basic`);
  await page.locator('body').click({ position: { x: 0, y: 0 } });

  // open stacking modals, close with escape key
  await page.click('[data-test-id="modal-stacking"]');
  await page.click('dialog[open] footer button');
  await takeScreenshot(page, `${componentName}-stacking`);
  await page.keyboard.press('Escape');
  await page.keyboard.press('Escape');

  // open modal with hidden close button, close with action button
  await page.click('button[data-test-id="modal-with-hidden-close-button"]');
  await takeScreenshot(page, `${componentName}-hidden-close-button`);
  await page.click('dialog[open] footer button');
  await takeScreenshot(page, `${componentName}-closed-modals`, { fullPage: true });

```

Only as an option. Depends on you

**Review by curdaj — APPROVED** [review]

**Comment by literat** [comment]
Maybe note to the future: Do we need to test those Modals visually? That is a lot of images. So wouldn't it be appropriate to test it by integration test style, e.g. click on a button, modal opens, find in DOM some nodes, click on a backdrop, click on the close button, fire ESC event, etc...

**Review by literat — APPROVED** [review]
I love the way you have refactored the test and created new helpers for it. Good Job 👍 

There is also an open question whether we should do these tests in a visual way. I can imagine that repeating this for Tooltip as well as for Dropdow will lead us to so many images.

**Comment by pavelklibani** [comment]
> Maybe note to the future: Do we need to test those Modals visually? That is a lot of images. So wouldn't it be appropriate to test it by integration test style, e.g. click on a button, modal opens, find in DOM some nodes, click on a backdrop, click on the close button, fire ESC event, etc...

We agreed to make tests for all modals in the office while you were in Poland. So I will leave it like this for now. We can remove some in the future.

**Comment by pavelklibani** [comment]
rebased to main

**Pull request merged**